New Feature: Sitewide Component Properties

HitmanInWis opened the pull request:
This is an **alpha** state feature (not yet vetted for production use) that HS2 Solutions put together at a company-wide 1 & 1/2 day hackathon.  We will have a PR for the documentation page as well, but wanted to get this up to see if people think this is a reasonable feature to add, or if AEM has a better way of doing this that we just couldn't find.

**Problem being solved:**
- Multiple instances of the same AEM component need to be configured separately, leading to:
  - Excessive copy/paste (error-prone)
  - No reuse of properties across instances without custom implementation
  - Labor-intensive and unmanageable for content changes
- Reference component solves some cases, but
  - Is an all or nothing deal - the reference is an exact copy of the parent
  - If not used carefully, Reference components can create a nasty spiderweb of content references across the site

**Goals of our solution:**
- No alteration of existing functionality - instance level component properties are unaffected
- New site-wide properties can be added to ensure consistency of a component's content across the entire site.  With site-wide properties, an update covers the entire site for that component (no copy and paste).
- Support overriding site-side properties by instance-level properties when set.
- Site-wide properties dialog has parity with OOTB instance-level properties dialog

**Installation:**
This feature is enabled by adding a cq:infoProvider to your page component, referencing com.adobe.acs.commons.wcm.impl.SitewideComponentPropsPageInfoProvider. See documentation at https://docs.adobe.com/docs/en/aem/6-1/develop/components/pageinfo.html for adding nodes to cq:infoProvider.

See our example implementation component at /apps/acs-commons/components/content/sitewide-component-props-example - the magic is added by the cq:dialogsitewide node which mimics the standard cq:dialog node.

**Still need to test/verify:**
- Ensure that properties can be sent out for translation
- Ensure that Reference component can work with global configs
- Ensure that path fields within global configs are updated if referenced paths are updated
- Ensure that merging of dialogsitewide.xml from resourceSuperType is working

**Future Features and Enhancements:**
- Only display the sitewide properties button for components that define a _cq_dialogsitewide node — Alternatively show a "no sitewide configs" message for components that do not have them
- Move the sitewide properties button on the authoring toolbar next to the wrench icon for the standard properties dialog
- Ensure authoring page refreshes when sitewide props are updated, since multiple component instances may be affected
- Flush dispatcher for all pages within the site using a component when site-wide properties are updated.
- Add support for site-wide properties that are not specific to a component type

justinedelson commented:
@HitmanInWis This is a really cool idea. I have a few quibbles with the implementation which I'll do via line comments, but I would strongly encourage you to continue to polish this and get it ready for inclusion. I think it would be a great addition to ACS AEM Commons.

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/impl/SitewideComponentPropsPageInfoProvider.java:
```diff
@@ -0,0 +1,46 @@
+ * PageInfoProvider which indicates that site-wide component properties
+ * are enabled.
+ */
+@Component
```
should probably have a REQUIRE configuration policy so deployers can opt-in to this feature.

HitmanInWis replied:
Developers already have to opt-in to use this by adding this class as one of the cq:infoProviders configured on their page component.  Do you feel the additional level of opt-in is necessary?

HitmanInWis replied:
Digging thru code, I notice that com.adobe.acs.commons.wcm.impl.WorkflowModelFilterPageInfoProvider, which implements PageInfoProvider, is currently configured w/o REQUIRE.  Is that provider fundamentally different than this one?  Or do you think we're fine w/o making this change?

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@ProviderType
+@Tag(bodyContentType = BodyContentType.JSP, value = "defineObjects")
+public class DefineObjects extends BodyTagSupport {
```
Please refactor this as a BindingsValuesProvider (which could, in turn, be used by a JSP tag) so that these merged properties can be used in Sightly and other scripting engines.

HitmanInWis replied:
This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+
+            // Build the path to the global config for this component
+            // <Lang Root>/jcr:content/sitewideprops/<component resource type>
+            String languageRoot = LanguageUtil.getLanguageRoot(currentNode.getPath());
```
Would it make sense to also support come kind of property to look up the site root? Not all AEM sites use language root-style paths.

HitmanInWis replied:
This is now refactored on the HS2 repo.  Once we finalize our updates I'll update this PR.

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+
+            // Send the Node back if it exist
+            if (r != null) {
+                JcrPropertyMap jpm = new JcrPropertyMap(r.adaptTo(Node.class));
```
I'm curious why this isn't just a ValueMap, obtained by calling r.getValueMap().

HitmanInWis replied:
We chose to use a JcrPropertyMap because the standard "properties" attribute in the pageContext which is a JcrPropertyMap instance.  We wanted to have the "shared" properties to be as effectively the same as instance properties as possible.  Using a ValueMap might not have a tangible difference, but unless there is a perceived benefit to doing so I'm inclined to keep as JcrPropertyMap for consistency w/OOTB AEM.

HitmanInWis replied:
Ignore that last comment :)

I had myself confused. This is now refactored on the HS2 repo.  Once we finalize our updates I'll update this PR.

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+        pageContext.setAttribute("mergedProperties", mergeProperties(localPropertyMap, sitewidePropertyMap));
+    }
+
+    private Map<String, Object> mergeProperties(JcrPropertyMap instanceProps, JcrPropertyMap sitewideProps) {
```
would be better to just use ValueMap here

HitmanInWis replied:
This is now refactored on the HS2 repo.  Once we finalize our updates I'll update this PR.

justinedelson commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private void setSitewideProperties() {
+        Node currentNode = (Node) this.pageContext.findAttribute("currentNode");
```
would prefer to use Resource rather than Node. We usually try to avoid using the JCR API where the Sling Resource API will work.

HitmanInWis replied:
This is now refactored on the HS2 repo.  Once we finalize our updates I'll update this PR.

justinedelson commented on content/src/main/content/META-INF/vault/filter.xml:
```diff
@@ -48,4 +48,5 @@
         <include pattern="/etc/acs-commons/workflow-remover(.*)"/>
     </filter>
     <filter root="/apps/cq/core/content/nav/tools/acs-commons"/>
+    <filter root="/apps/cq/gui/components/authoring/dialogsitewide"/>
```
does this need to be in /apps/cq?

HitmanInWis replied:
Moved to /apps/acs-commons

This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

justinedelson commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/sitewide-component-props/sitewide-component-props.js:
```diff
@@ -0,0 +1,51 @@
+                                    editable.type;
+            
+            editable.config.dialogSrc = siteWideDialogSrc;
+            editable.config.dialog = editable.config.dialog.replace("cq:dialog", "cq:dialogsitewide");
```
can we use acs-commons:dialogsitewide instead of cq:dialogsitewide? In theory, only AEM should define node or property names using the 'cq' namespace.

davidjgonzalez replied:
this would be the first introduction of the acs-commons namespace?

justinedelson replied:
it would be and we should be thoughtful about that step, *but* as I said, introducing new node names in the `cq` namespace is wrong.

davidjgonzalez replied:
Not arguing that - but why not give it some none namespaced name, we do that fairly often?

HitmanInWis replied:
Moved cq:dialogshared nodes to dialogshared (no namespace).

This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

justinedelson commented on content/src/main/content/jcr_root/apps/cq/gui/components/authoring/dialogsitewide/dialogsitewide.jsp:
```diff
@@ -0,0 +1,17 @@
+
+    // Ensure the path exists in the JCR so that we don't get a NPE
+    // when attempting to load the dialog.
+    JcrUtil.createPath(dialogSitewideDataPath, "nt:unstructured", "nt:unstructured", dialogSitewideSession, false);
```
This is fairly problematic as presumably not all users will have write access to the sitewide data path. Would IMHO be worth investigating the case where write permissions aren't granted.

justinedelson commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/sitewide-component-props/sitewide-component-props.js:
```diff
@@ -0,0 +1,51 @@
+            var langRootRegexp = /^(\/content\/([^/]+\/)+([a-zA-Z]{2}([_-][a-zA-Z]{2})?))(|\/.*)$/g;
+            var match = langRootRegexp.exec(dialogSrcArray[1]);
+            
+            var siteWideDialogSrc = dialogSrcArray[0].replace("_cq_dialog", "_cq_dialogsitewide") +
```
same comment as below about the node name.

HitmanInWis replied:
This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/sitewide-component-props/.content.xml:
```diff
@@ -0,0 +1,6 @@
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:sling="http://sling.apache.org/jcr/sling/1.0"
+      jcr:primaryType="cq:ClientLibraryFolder"
+      categories="[cq.authoring.editor]"
+      dependencies="[granite.shared,granite.ui.foundation.admin,rte.coralui2,coralui2.imageeditor,history,lawnchair,granite.ui.notifications,granite.ui.annotations,cq.siteadmin.common.notifications,cq.social.commons]"/>
```
Does this require all these dependencies?

HitmanInWis replied:
This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-commons/components/content/sitewide-component-props-example/.content.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Could we move the example impl into the docs or ACS Samples project (and link from docs)?

Apologies if I missed this being addressed already

HitmanInWis replied:
@davidjgonzalez Looking at the ACS Samples project, wondering the best way to incorporate this.  There's a site at /content/acs-samples but it contains a (currently broken on 6.1) sample angularJS site called "AngularJS App" directly below it, with /content/acs-samples redirecting to it at /content/acs-samples/angularjs-app.

Would it make sense for me to add sample pages to the same angularjs-app website that demonstrate this new feature?  Or should I create a new site at /content/acs-samples/acs-shared-props?  Want to make sure I'm in line with the vision for that project.

davidjgonzalez commented on bundle/src/main/java/com/adobe/acs/commons/wcm/tags/DefineObjects.java:
```diff
@@ -0,0 +1,101 @@
+
+        // Add Component Global Configs
+        if (sitewideProps != null) {
+            Set<String> sitewideKeys = sitewideProps.keySet();
```
mergedProprties.putAll(..) if you're editing anyhow?

davidjgonzalez commented:
Very nice! I think this could be handy for sections of sites as well.

Following JE's idea of allowing the page w the shared properties to be specified in other ways than the language util... Wdyt about

1. Adding ability to select a parent node via looking up the tree 
2. Exposing this "src" page via the page info provider and letting the JS be told the location (consolidate the logic server side)
3. Changing name to be more general - maybe SharedComponentProperties (or something like that?)

I think other than the naming the rest could be v2 additions as well.

HitmanInWis commented:
Justin and David, thanks for the immediate feedback. I was on out of office all week, but HS2 Solutions will take a look at all of your comments and address and/or respond to them.

brian-langeland commented:
Thank you for all the feedback. The HS2 Solutions team will be picking this back up and working through the changes requested, and should have some updates to this PR ready in the next couple of weeks.

HitmanInWis commented:
@davidjgonzalez regarding your question of making it usable for sections of the site:
1) Updated to specify a home page path pattern via OSGi configs.
2) Done.  Home page will come from the PageInfoProvider and be used by the JS
3) Updated the naming of everything from "sitewide" to "shared"

This doesn't yet add support for shared properties in "sections" of the site but sets us up for it if someone wants to add in the future.

This is now refactored on the HS2 repo. Once we finalize our updates I'll update this PR.

HitmanInWis commented:
Is there a way for me to augment the component edit configs printed to the HTML behind the scenes for the authoring UI.  The configs I'm talking about look like this in the Touch UI:

`<cq data-path="/content/mysite/mypage/jcr:content/main-par/mycomponent" data-config="{"path":"/content/mysite/mypage/jcr:content/main-par/mycomponent","slingPath":"/content/mysite/mypage/jcr:content/main-par/mycomponent.html","dialog":"/apps/mysite/components/content/mycomponent/cq:dialog","dialogLoadingMode":"auto","dialogLayout":"auto","dialogSrc":"/apps/mysite/components/content/mycomponent/_cq_dialog.html/content/mysite/mypage/jcr:content/main-par/mycomponent","dialogClassic":"/apps/mysite/components/content/mycomponent/dialog","type":"mysite/components/content/mycomponent","csp":"freeform|basepage|page/main-par|responsivegrid/mycomponent"}">`

What I'm looking to do is update the "data-config" value, adding extra values that I can then read from JavaScript similar to how "dialogSrc" can be read with `editable.config.dialogSrc` in JS.  That way I can add something like a "dialogSharedSrc" value that the JS can key off of to know if the shared props dialog is available for a given component.

HitmanInWis commented:
Ok, PR has been updated in response to all of the feedback.  This feature is now in a much better place than it was after our initial 1.5 days of effort :).

Hoping to have a PR up for the documentation (including the sample component) tomorrow.

Issues resolved from initial PR:
- All requested code updates/refactors
- Added Classic UI support
- Moved the sitewide properties button on the authoring toolbar next to the wrench icon for the standard properties dialog
- Added support for shared properties that are not specific to a component type (i.e. global properties)

Verified all of the following are working (initial PR we had not yet tested):
- Translation includes shared properties
- Path fields in shared properties are automatically updated when their referenced path is updated
- Touch UI dialogs are merged from ancestor resource types
- Reference component works fine with referencing a component using shared properties

Outstanding Issues:
- Attempted to determine a way to show the shared properties buttons only when dialogs exist
    - This is the main issue I'd like to see resolved.  If anyone has any suggestions, please let me know.
    - I have an open issue on Adobe forums to determine a solution: http://help-forums.adobe.com/content/adobeforums/en/experience-manager-forum/adobe-experience-manager.topic.html/forum__tjvx-is_there_a_wayform.html
- Ensure authoring page refreshes when sitewide props are updated, since multiple component instances may be affected
    - Not sure if this one is worth pursuing - the OOTB AEM Reference component does not do this (i.e. the author must refresh the page to see an update reflected in a Reference component that references a component on the same page)
- Flush dispatcher for all pages within the site using a component when site-wide properties are updated.
    - If this is added, it would need some level of configuration because blowing the entire dispatcher cache away for a single update might be a bit heavy-handed for a high traffic site.

HitmanInWis commented on bundle/src/main/java/com/adobe/acs/commons/wcm/impl/SharedComponentPropertiesBindingsValuesProvider.java:
```diff
@@ -0,0 +1,108 @@
+ * to instance-level values, then shared values, and finally global
+ * values when properties exist at multiple levels with the same name.
+ */
+@org.apache.felix.scr.annotations.Component(
```
Should this have a configuration policy of REQUIRE?

davidjgonzalez replied:
This does not require a configuration, but only because it depends on your PageRootProvider to be satisfied.

HitmanInWis replied:
True, but PageRootProvider was created generically and could technically be reused by other features or even custom code in the future.

HitmanInWis commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/shared-component-properties-classic/shared-component-properties-classic.js:
```diff
@@ -0,0 +1,105 @@
+                // Using 'evt.ownerCt.defaults.scope' instead of 'this' because
+                // 'this' has a permanent reference to the first component for
+                // which the menu is launched.
+                editSharedHandler.call(evt.ownerCt.defaults.scope, "shared", CQ.wcm.EditBase.EDITSHARED);
```
Really crappy scoping issue here where I couldnt use `this` because of how ActionsConvertors are singletons.

HitmanInWis commented:
Documentation PR is up for internal review at HS2.  Can be seen at https://github.com/HS2-SOLUTIONS/adobe-consulting-services.github.io/pull/1/files

HitmanInWis commented on content/src/main/content/jcr_root/apps/acs-commons/gui/components/authoring/dialogshared/dialogshared.jsp:
```diff
@@ -0,0 +1,20 @@
+    // Ensure the path exists in the JCR so that we don't get a NPE
+    // when attempting to load the dialog.
+    if (!dialogSharedSession.nodeExists(dialogSharedDataPath)) {
+        JcrUtil.createPath(dialogSharedDataPath, "nt:unstructured", "nt:unstructured", dialogSharedSession, false);
```
The issue is that the OOTB AEM dialog code pukes if the node is multiple levels deep and the nodes in between dont yet exist.

SharedComponentPropsPageInfoProvider checks if the current user has permissions to write that node, and if not then flags the shared component properties feature as disabled, which in turn makes it so the UI does not give the user the buttons to configure shared/global configs (and thus they will not hit this code).

HitmanInWis commented:
Official Documentation PR is up now: https://github.com/Adobe-Consulting-Services/adobe-consulting-services.github.io/pull/84

HitmanInWis commented:
Is there anything more I can do to help move this PR forward?  Looking forward to using this feature on projects w/o having to pull in all of the code manually.  I understand the folks that maintain this repo are very busy, so by no means am I pushing - just want to make sure I'm doing whatever I can to move this forward as I feel this feature is something a lot of folks could leverage.

davidjgonzalez commented:
@HitmanInWis im going to try to take a look at this week/weekend -- been pretty busy w/ work.

Cursory glance through, nothing jumps out as problematic, but let try it out as its a larger feature.

Thanks for the submission and patience!

HitmanInWis commented:
Thanks a lot David, we appreciate it and look forward to your feedback!

HitmanInWis commented:
@davidjgonzalez Pulled in your changes - thanks for cleaning things up a bit!

I did reverse the modification of the regexp as it makes the match too loose in my opinion.  Let me know what you think - we can easily revert my last commit if you disagree.

davidjgonzalez commented:
@HitmanInWis gotcha on the regex; lets leave it as youve adjusted.

davidjgonzalez commented:
@HitmanInWis fyi - this is in the latest ACS Commons release (2.8.2/3.5.0) -- thanks again for this feature. very slick.

HitmanInWis commented:
@davidjgonzalez Awesome, great to hear!  When do you anticipate the documentation PR (https://github.com/Adobe-Consulting-Services/adobe-consulting-services.github.io/pull/84) will go through so that people can see the new feature and how to enable it on the main website?

davidjgonzalez commented:
docs are up!
